Liveliness QoS implementation

dennis-adlink opened the pull request:
In this PR, support is added for the liveliness QoS with automatic and manual-by-participant liveliness kinds. It also adds support for setting and requesting the lease_duration for a writer using this QoS. 
In the Cyclone api the function _dds_assert_liveliness_ is added to assert liveliness on a participant, which can be used when any writer is given liveliness kind manual-by-participant. 
Liveliness kind manual-by-topic is not yet supported. Although I'm planning to add this in the near future, I think it's useful to have this PR reviewed at this point. 

Signed-off-by: Dennis Potman <dennis.potman@adlinktech.com>

eboasson commented:
@dennis-adlink with apologies for the long delay in responding: thank you very much for contributing this. It looks good to me, but there are some bits and pieces that I want to consider a bit longer before accepting the PR. Not because of what I see, but because of the potential for race conditions and general nastiness arising out of unfortunate timings. I hope you don't mind my taking a bit more time.

dennis-adlink commented:
In the last 5 commits in this PR I have addressed the issues from the review and added some additional tests. The commit messages give some details on what I've changed, but generally I agreed with the review comments and followed these suggestions. The liveliness kind manual-by-topic is not implemented yet, but I think this will not be much work now I've implemented the other two kinds. I will create a separate PR for this.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -1463,14 +1488,14 @@ static void writer_drop_local_connection (const struct ddsi_guid *wr_guid, struc
 }
 
-static void reader_drop_connection (const struct ddsi_guid *rd_guid, const struct proxy_writer *pwr)
+static void reader_drop_connection (const struct ddsi_guid *rd_guid, const struct proxy_writer *pwr, bool unmatch)
```
Replying to myself after considering the two points I raised: that race between transitions from alive to not-alive and vice-versa cannot exist because the proxy writer lock is held for the duration of the transition.

Holding the proxy writer's lock while calling listeners is not ideal, but that same problem already exists for other listeners. The argument for invoking listeners inline is that it gives the lowest possible latency as well as a number of guarantees you cannot otherwise have, and the price you pay is that you have to be careful what you do in a listener. This is not something to make a fuss over that here & now.

The other point, delete/loss-of-liveliness in parallel I haven't yet tried to figure out in detail. But I have decided that will not block acceptance of the pull request.

dennis-adlink commented:
@eboasson In the two additional commits on this PR, I've added the implementation for liveliness kind manual-by-topic and processed your review comments on my earlier commits. For a (brief) description of the changed I've made, see the commit messages.

dennis-adlink commented:
@eboasson The Travis build result for build [611948174](https://travis-ci.org/eclipse-cyclonedds/cyclonedds/jobs/611948174) is obviously telling me something is still wrong with locking when a proxy writer is deleted, so I'll analyze and fix this (so please ignore the re-review request until this is fixed).

eboasson commented:
@dennis-adlink the domain id/tag feature that just got pushed onto master requires a change to the liveliness tests (because now the two domains you create will no longer communicate even when they are on the same port number unless you force the advertised domain id to match).

```patch
diff --git a/src/core/ddsc/tests/liveliness.c b/src/core/ddsc/tests/liveliness.c
index 7063410b..ad0e9d29 100644
--- a/src/core/ddsc/tests/liveliness.c
+++ b/src/core/ddsc/tests/liveliness.c
@@ -30,8 +30,8 @@
 
 #define DDS_DOMAINID_PUB 0
 #define DDS_DOMAINID_SUB 1
-#define DDS_CONFIG_NO_PORT_GAIN "${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}<Discovery><Ports><DomainGain>0</DomainGain></Ports></Discovery>"
-#define DDS_CONFIG_NO_PORT_GAIN_LOG "${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}<Tracing><OutputFile>cyclonedds_liveliness_tests.${CYCLONEDDS_DOMAIN_ID}.${CYCLONEDDS_PID}.log</OutputFile><Verbosity>finest</Verbosity></Tracing><Discovery><Ports><DomainGain>0</DomainGain></Ports></Discovery>"
+#define DDS_CONFIG_NO_PORT_GAIN "${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}<Discovery><ExternalDomainId>0</ExternalDomainId></Discovery>"
+#define DDS_CONFIG_NO_PORT_GAIN_LOG "${CYCLONEDDS_URI}${CYCLONEDDS_URI:+,}<Tracing><OutputFile>cyclonedds_liveliness_tests.${CYCLONEDDS_DOMAIN_ID}.${CYCLONEDDS_PID}.log</OutputFile><Verbosity>finest</Verbosity></Tracing><Discovery><ExternalDomainId>0</ExternalDomainId></Discovery>"
 
 uint32_t g_topic_nr = 0;
 static dds_entity_t g_pub_domain = 0;
```

dennis-adlink commented:
@eboasson Thanks for your help wrt correctly handling the liveliness changed event! I have cherry picked your commits in this branch. I've added a fix that prevents the proxy writer from getting alive (via `handle_regular` when data is received) after `delete_proxy_writer` scheduled the gc `gc_delete_proxy_writer` for cleaning up. 

In addition I have changed the behaviour of the counters for liveliness events a bit, so that the `alive_count_change` and `not_alive_count_change` fields reflect the actual change (the net result of the increase and decrease of the count) of the `alive_count`/`not_alive_count` fields and not only the increase. I think this is closer to what the DDS specification describes (at least, that is my interpretation, see the table at the end of section 2.2.4.1). I've also changed the tests accordingly.

eboasson commented:
Thanks @dennis-adlink !

I've rebased it to deal with some changes I merged earlier and squashed some of the commits together (the "fixup" kind, in particular) and merged it as PR #342.
